File: php/src/Domain/Reporting/ReportGenerator.php
<?php

namespace Katalogs\BookInvoicing\Domain\Reporting;

use Katalogs\BookInvoicing\Domain\Finance\CurrencyConverter;
use Katalogs\BookInvoicing\Domain\Purchase\Invoice;
use Katalogs\BookInvoicing\Domain\Purchase\MainRepository;
use Katalogs\BookInvoicing\Domain\Purchase\PurchasedBook;
use Katalogs\BookInvoicing\Domain\Purchase\Repository;

class ReportGenerator
{
    private readonly Repository $repository;

    public function __construct()
    {
        $this->repository = MainRepository::configuredRepository();
    }

    public function getTotalAmount(): float {
        $invoices = $this->repository->getInvoices();
        $totalAmount = 0.0;

        /** @var Invoice $invoice */
        foreach ($invoices as $invoice) {
            $totalAmount += CurrencyConverter::toUSD(
                $invoice->computeTotalAmount(),
                $invoice->country->currency
            );
        }
        return $this->getRoundedAmount($totalAmount);
    }

    private function getRoundedAmount(float $totalAmount): float
    {
        return round($totalAmount * 100) / 100;
    }

    public function getTotalSoldBooks(): float {
        $invoices = $this->repository->getInvoices();
        $totalSoldBooks = 0;

        /** @var Invoice $invoice */
        foreach ($invoices as $invoice) {
            /** @var PurchasedBook $purchasedBook */
            foreach ($invoice->getPurchasedBooks() as $purchasedBook) {
                $totalSoldBooks = $totalSoldBooks + $purchasedBook->quantity;
            }
        }
        return $totalSoldBooks;
    }

    public function getNumberOfIssuedInvoices(): int {
        $invoices = $this->repository->getInvoices();
        return count($invoices);
    }
}

File: java/src/main/java/org/peaksys/gameoftechs/report/ReportGenerator.java
package org.peaksys.gameoftechs.report;

import org.peaksys.gameoftechs.MainRepository;
import org.peaksys.gameoftechs.finance.CurrencyConverter;
import org.peaksys.gameoftechs.purchase.Invoice;
import org.peaksys.gameoftechs.purchase.PurchasedBook;
import org.peaksys.gameoftechs.storage.Repository;

import java.util.List;
import java.util.Map;

public class ReportGenerator {

    private final Repository repository = MainRepository.configuredRepository();

    public double getTotalAmount() {
        Map<Integer, Invoice> invoiceMap = repository.getInvoiceMap();

        var totalAmount = invoiceMap
                .values()
                .stream()
                .mapToDouble(invoice -> CurrencyConverter.toUSD(invoice.computeTotalAmount(), invoice.getCountry().currency()))
                .sum();
        return getRoundedAmount(totalAmount);
    }

    private double getRoundedAmount(double totalAmount) {
        return Math.round(totalAmount * 100.0) / 100.0;
    }

    public int getTotalSoldBooks() {
        Map<Integer, Invoice> invoiceMap = repository.getInvoiceMap();
        return invoiceMap
                .values()
                .stream()
                .flatMap(invoice -> invoice.getPurchasedBooks().stream())
                .mapToInt(PurchasedBook::quantity)
                .sum();
    }

    public long getNumberOfIssuedInvoices() {
        Map<Integer, Invoice> invoiceMap = repository.getInvoiceMap();
        return invoiceMap.values().size();
    }
}
